src/runtime/HalideRuntime.h

```diff
@@ -541,6 +546,38 @@ extern int halide_copy_to_host(void *user_context, struct halide_buffer_t *buf);
 extern int halide_copy_to_device(void *user_context, struct halide_buffer_t *buf,
                                  const struct halide_device_interface_t *device_interface);
 
+/** Copy data from one buffer to another. The buffers may have
+ * different shapes and sizes, but the destination buffer's shape must
+ * be contained within the source buffer's shape. The source data is
```

Comment:
We should explicitly spell out the preconditions and postconditions on src and dst state. The completely general API allows the initial state to be almost anything. The src can be on any or no device and if it does have a device, may have a null host pointer. Basically the only requirement would be that at least one thing is non-null. (Likely with the caveat that a zero extent copy is a nop no matter what.) The dirty bits can be any of the three valid states, and the dst can be on any device or no device and can probably even have a null host pointer, in which case we allocate one if dst_device_interface is null. The latter requires the caller to call halide_free on the host pointer if it is allocated. The final state of the dirty bits would also be somewhat wide open, though some specification can be made.

In the above, the only thing that I don't see being easily doable is a combined device and host allocation as there is no way to control whether one wants one or not. We could assume one always does if it is available and the dst buffer does not already have some sort of allocation. Note that specifying all of this is important because the caller will have to make sepcific deallocation calls to cleanup the result.

All of that is both a bit of work to specify and will absolutely require test coverage as the internals are non-trivial.

Perhaps the most restricted interface insists that the source and destination allocations already exist and the other side of each buffer is not dirty. In this case, one would probably just have a bool to indicate whether copying to the host or device side of the destination.

I am aware that this borders on being pedantic, but I really do feel that if we aren't rigorous about specifying this call, it will cause problems.